File: ik_problem/sorting/3_sum.py
"""
solution: http://www.programcreek.com/2012/12/leetcode-3sum/
"""
def three_sum(a_list):
    res = []

    if len(a_list) < 3:
        return res

    sorted_array = sorted(a_list)

    i = 0
    while i < len(sorted_array) - 2:
        if i == 0 or sorted_array[i] > sorted_array[i - 1]:
            negate = -sorted_array[i]

            start = i + 1
            end = len(sorted_array) - 1

            while start < end:
                if sorted_array[start] + sorted_array[end] == negate:
                    temp = []
                    temp.append(sorted_array[end])
                    temp.append(sorted_array[start])
                    temp.append(sorted_array[i])

                    res.append(temp)
                    start += 1
                    end -= 1

                    # avoid duplicate solutions
                    while (start < end and sorted_array[end] == sorted_array[end + 1]):
                        end -= 1

                    while start < end and sorted_array[start] == sorted_array[start - 1]:
                        start -= 1

                elif sorted_array[start] + sorted_array[end] < negate:
                    start += 1
                else:
                    end -= 1

        i += 1

    return res

print three_sum([-1, 0, 1, 2, -1, -4])

File: leetcode/array/ruby/3_sum.rb
=begin
https://leetcode.com/problems/3sum/

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]

algorithm: The idea is to sort an input array and then run through all indices of a possible first element of a triplet. For each possible first element we make a standard bi-directional 2Sum sweep of the remaining part of the array. Also we want to skip equal elements to avoid duplicates in the answer without making a set or smth like that.

O(n^2) time O(n) space


# https://discuss.leetcode.com/topic/22619/python-easy-to-understand-solution-o-n-n-time

In the inner while loop why do we have to  do l += 1; r -= 1 after skipping duplicates?

If we didn't skip any duplicates, then we definitely have to move r and l.

If we skipped duplicates, why do we still have to move?

Imagine nums = [-4, -1, -1 ... 5]:

i is the element -4
l is the element -1
r is the element 5
s = 0

At this point the first l is part of the result we want and we've skipped l to the last -1 (using the first inner while loop: line 62). In the next iteration we still don't want to include -1 in the result (it'll result in a duplicate). So we need to move l again, and also move r as usual. That's why!

The same explanation works for r
=end
def three_sum(nums)
  result = []
  nums.sort!

  (nums.size - 2).times do |i|
    next if i > 0 && nums[i] == nums[i - 1]

    l = i + 1
    r = nums.size - 1

    while l < r
      total = nums[i] + nums[l] + nums[r]

      if total < 0
        l += 1
      elsif total > 0
        r -= 1
      else
        result << [nums[i], nums[l], nums[r]]

        l += 1 while l < r && nums[l] == nums[l + 1]

        r -= 1 while l < r && nums[r] == nums[r - 1]

        l += 1
        r -= 1
      end
    end
  end

  result
end

if $PROGRAM_NAME == __FILE__
  p three_sum([-1, 0, 1, 2, -1, -4])
  p three_sum([0, 0, 0, 0, 0, 0])
end
